File: client/EditConnectionWindow.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace repatriator_client
{
    public partial class EditConnectionWindow : Form
    {
        private enum Mode
        {
            NewMode,
            EditMode,
        }

        private Mode mode;

        private Connection connection;

        private const string static_password = "12345678";
           
        public EditConnectionWindow()
        {
            InitializeComponent();
        }

        public Connection showNew(IWin32Window owner)
        {
            this.Text = "New Connection - Repatriator";
            mode = Mode.NewMode;

            // clear controls
            urlTextBox.Text = "";
            portTextBox.Text = "";
            userNameTextBox.Text = "";
            savePasswordCheckBox.Checked = false;
            passwordTextBox.Text = "";

            enableCorrectControls();

            connection = new Connection();
            this.ShowDialog(owner);
            return connection;
        }

        public Connection showEdit(IWin32Window owner, Connection conn)
        {
            this.Text = "Edit Connection - Repatriator";
            mode = Mode.EditMode;

            // load values from connection
            urlTextBox.Text = conn.url;
            portTextBox.Text = conn.port.ToString();
            userNameTextBox.Text = conn.username;
            bool save_password = conn.password.Length != 0;
            savePasswordCheckBox.Checked = save_password;
            passwordTextBox.Text = save_password ? static_password : "";

            enableCorrectControls();

            connection = conn;
            this.ShowDialog(owner);

            return connection;
        }

        private void updateConnectionWithControls()
        {
            connection.url = urlTextBox.Text;
            connection.port = 0;
            int.TryParse(portTextBox.Text, out connection.port);
            connection.username = userNameTextBox.Text;
            connection.password = (savePasswordCheckBox.Checked && passwordTextBox.Text != static_password) ? passwordTextBox.Text : "";
        }

        private void enableCorrectControls()
        {
            passwordTextBox.Enabled = savePasswordCheckBox.Checked;
            passwordLabel.Enabled = savePasswordCheckBox.Checked;
        }

        private void savePasswordCheckBox_CheckedChanged(object sender, EventArgs e)
        {
            enableCorrectControls();
        }

        private void okButton_Click(object sender, EventArgs e)
        {
            updateConnectionWithControls();
            this.Close();
        }

        private void cancelButton_Click(object sender, EventArgs e)
        {
            connection = null;
            this.Close();
        }

        private void passwordTextBox_Enter(object sender, EventArgs e)
        {
            if (passwordTextBox.Text == static_password)
                passwordTextBox.Text = "";
        }

        private void portTextBox_KeyPress(object sender, KeyPressEventArgs e)
        {
            if (!(('0' <= e.KeyChar && e.KeyChar <= '9') || e.KeyChar == '\b'))
                e.Handled = true;
        }
    }
}

File: client/EditConnectionWindow.cpp
#include "EditConnectionWindow.h"
#include "ui_EditConnectionWindow.h"

#include "ConnectionSettings.h"

#include <QApplication>
#include <QWidget>

EditConnectionWindow * EditConnectionWindow::s_instance = NULL;
const char * EditConnectionWindow::c_static_password = "12345678";

EditConnectionWindow::EditConnectionWindow(QWidget *parent) :
    QDialog(parent),
    ui(new Ui::EditConnectionWindow)
{
    ui->setupUi(this);
    bool success;
    success = connect(QApplication::instance(), SIGNAL(focusChanged(QWidget*,QWidget*)), this, SLOT(delegateFocusEvent(QWidget*,QWidget*)));
    Q_ASSERT(success);
    success = connect(this, SIGNAL(accepted()), this, SLOT(handleAccepted()));
    Q_ASSERT(success);
    success = connect(this, SIGNAL(rejected()), this, SLOT(handleRejected()));
    Q_ASSERT(success);
}

EditConnectionWindow::~EditConnectionWindow()
{
    delete ui;
}

void EditConnectionWindow::changeEvent(QEvent *e)
{
    QDialog::changeEvent(e);
    switch (e->type()) {
    case QEvent::LanguageChange:
        ui->retranslateUi(this);
        break;
    default:
        break;
    }
}

EditConnectionWindow * EditConnectionWindow::instance()
{
    if (! s_instance)
        s_instance = new EditConnectionWindow();
    return s_instance;
}

bool EditConnectionWindow::showNew(ConnectionSettings * connection)
{
    this->setWindowTitle(tr("New Connection - Repatriator"));
    m_mode = NewMode;
    m_connection = connection;

    // clear controls
    ui->addressLineEdit->setText("");
    ui->portLineEdit->setText("");
    ui->usernameLineEdit->setText("");
    ui->savePasswordCheckBox->setChecked(false);
    ui->passwordLineEdit->setText("");

    ui->addressLineEdit->setFocus(Qt::OtherFocusReason);

    enableCorrectControls();

    this->exec();
    return m_accepted;
}

bool EditConnectionWindow::showEdit(ConnectionSettings * connection)
{
    this->setWindowTitle(tr("Edit Connection - Repatriator"));
    m_mode = EditMode;
    m_connection = connection;

    // load values from connection
    ui->addressLineEdit->setText(connection->host);
    ui->portLineEdit->setText(QString::number(connection->port));
    ui->usernameLineEdit->setText(connection->username);
    bool save_password = connection->password.length() != 0;
    ui->savePasswordCheckBox->setChecked(save_password);
    ui->passwordLineEdit->setText(save_password ? c_static_password : "");

    ui->addressLineEdit->setFocus(Qt::OtherFocusReason);

    enableCorrectControls();

    this->exec();
    return m_accepted;
}

void EditConnectionWindow::enableCorrectControls()
{
    ui->passwordLineEdit->setEnabled(ui->savePasswordCheckBox->isChecked());
    ui->passwordLabel->setEnabled(ui->savePasswordCheckBox->isChecked());
}
void EditConnectionWindow::handleAccepted()
{
    updateConnectionWithControls();
    m_accepted = true;
}

void EditConnectionWindow::handleRejected()
{
    m_accepted = false;
}

void EditConnectionWindow::updateConnectionWithControls()
{
    m_connection->host = ui->addressLineEdit->text();
    m_connection->port = ui->portLineEdit->text().toInt();
    m_connection->username = ui->usernameLineEdit->text();
    if (ui->savePasswordCheckBox->isChecked() && ui->passwordLineEdit->text() != c_static_password)
        m_connection->password = ui->passwordLineEdit->text();
    else
        m_connection->password = QString();
}

void EditConnectionWindow::on_savePasswordCheckBox_toggled(bool checked)
{
    Q_UNUSED(checked);
    enableCorrectControls();
}

void EditConnectionWindow::delegateFocusEvent(QWidget *from, QWidget *to)
{
    Q_UNUSED(from);
    if (to == ui->passwordLineEdit)
        passwordLineEdit_gotFocus();
}

void EditConnectionWindow::passwordLineEdit_gotFocus()
{
    if (ui->passwordLineEdit->text() == c_static_password)
        ui->passwordLineEdit->setText(c_static_password);
}
